mayastor/src/bdev/nexus/nexus_persistence.rs

```diff
@@ -0,0 +1,146 @@
+use crate::{
+    bdev::{
+        nexus::nexus_child::ChildState::Faulted,
+        ChildState,
+        Nexus,
+        Reason,
+    },
+    persistent_store::PersistentStore,
+    sleep::mayastor_sleep,
+};
+use rpc::persistence::{
+    ChildInfo,
+    ChildState as PersistentChildState,
+    NexusInfo,
+    Reason as PersistentReason,
+};
+use std::time::Duration;
+
+type ChildUuid = String;
+
+/// Defines the type of persist operations.
+pub(crate) enum PersistOp {
+    /// Create a persistent entry.
+    Create,
+    /// Update a persistent entry.
+    Update((ChildUuid, ChildState)),
+    /// Save the clean shutdown variable.
+    Shutdown,
+}
+
+impl Nexus {
+    /// Persist information to the store.
+    pub(crate) async fn persist(&self, op: PersistOp) {
+        if !PersistentStore::enabled() {
+            return;
+        }
+        let mut nexus_info = self.nexus_info.lock().await;
+        match op {
+            PersistOp::Create => {
+                // Initialisation of the persistent info will overwrite any
+                // existing entries.
+                // This should only be called on nexus creation.
+                nexus_info.clean_shutdown = false;
+                self.children.iter().for_each(|c| {
+                    let state: PersistentChildState = c.state().into();
+                    let reason: PersistentReason = c.state().into();
+                    nexus_info.children.push(ChildInfo {
+                        uuid: c.uuid(),
+                        state: state as i32,
+                        reason: reason as i32,
+                    });
+                });
+            }
+            PersistOp::Update((uuid, state)) => {
+                // Only update the state of the child that has changed. Do not
+                // update the other children or "clean shutdown" information.
+                // This should only be called on a child state change.
+                nexus_info.children.iter_mut().for_each(|c| {
+                    let persistent_state: PersistentChildState = state.into();
```

Comment:
Wouldn't it be better to move 'let persistent_state' and 'let persistent_reason' under 'if c.uuid == uuid {' and instantiate them only when target replica is found ?

Reply:
Good spot, thanks. Done.